### PR TITLE
feat: add activity forms

### LINKED ISSUE #434: Add "new" and "edit" activity pages

### PR DESCRIPTION
**Pull request by danielsp45**

### CODE REVIEW

File: assets/css/app.css [hunk 1]
```diff
@@ -1,3 +1,3 @@
 @import "tailwindcss/base";
 @import "tailwindcss/components";
-@import "tailwindcss/utilities";
\ No newline at end of file
+@import "tailwindcss/utilities";
```
Comment (ruilopesm) [inline assets/css/app.css]: `git checkout develop -- assets/css/app.css`

File: lib/atomic_web/live/activity_live/form_component.ex [hunk 2]
```diff
@@ -4,23 +4,28 @@ defmodule AtomicWeb.ActivityLive.FormComponent do
   alias Atomic.Activities
   alias Atomic.Departments
 
+  @extensions_whitelist ~w(.jpg .jpeg .gif .png)
```
Comment (ruilopesm) [inline lib/atomic_web/live/activity_live/form_component.ex]:
```suggestion
```

File: lib/atomic_web/live/activity_live/form_component.ex [hunk 3]
```diff
@@ -4,23 +4,28 @@ defmodule AtomicWeb.ActivityLive.FormComponent do
...
+
   @impl true
   def mount(socket) do
-    {:ok, socket}
+    {:ok,
+     socket
+     |> allow_upload(:image, accept: @extensions_whitelist, max_entries: 1)}
```
Comment (ruilopesm) [inline lib/atomic_web/live/activity_live/form_component.ex]:
```suggestion
     |> allow_upload(:image, accept: Atomic.Uploader.extensions_whitelist(), max_entries: 1)}
```

File: lib/atomic_web/live/activity_live/form_component.html.heex [hunk 4]
```diff
@@ -1,51 +1,262 @@
...
+
+        <div class="flex flex-row mt-4 flex-col md:flex-row justify-center md:justify-start">
+          <div class="flex flex-col gap-y-1">
+            <div x-data="{
+        open: false,
+        toggle() { this.open = this.open ? this.close() : true },
+        close() { this.open = false },
+      }" @keydown.escape.prevent.stop="close()" @focusin.window="! $refs.panel.contains($event.target) && close()">
+              <span class="font-sm pl-1 text-base text-gray-800 sm:text-lg">
+                Departments
```
Comment (ruilopesm) [inline lib/atomic_web/live/activity_live/form_component.html.heex]: Activities don't depend on departments anymore.

File: lib/atomic_web/live/activity_live/index.ex [hunk 5]
```diff
@@ -28,6 +28,7 @@ defmodule AtomicWeb.ActivityLive.Index do
      socket
      |> assign(:page_title, gettext("Activities"))
      |> assign(:current_page, :activities)
+     |> assign(:current_user, socket.assigns.current_user)
```
Comment (ruilopesm) [inline lib/atomic_web/live/activity_live/index.ex]: Why is this assign needed? The `current_user` is injected in all live views, by a hook present at the router level.

File: lib/atomic_web/live/activity_live/index.html.heex [hunk 6]
```diff
@@ -47,7 +47,7 @@
     <!-- When there aren't any activities -->
     <%= if @empty? and @has_permissions? do %>
       <div class="mt-32">
-        <.empty_state id="empty_state" url={Routes.activity_new_path(@socket, :new, @current_organization)} plural_placeholder="activities" placeholder="activity" />
+        <.empty_state id="empty_state" url={Routes.activity_new_path(@socket, :new, @current_user.current_organization)} plural_placeholder="activities" placeholder="activity" />
```
Comment (ruilopesm) [inline lib/atomic_web/live/activity_live/index.html.heex]: `@current_organization` should be enough.

File: lib/atomic_web/live/activity_live/new.ex [hunk 7]
```diff
@@ -22,9 +23,12 @@ defmodule AtomicWeb.ActivityLive.New do
       }
     ]
 
+    organization = Organizations.get_organization!(organization_id)
+
     {:noreply,
      socket
      |> assign(:page_title, gettext("New Activity"))
+     |> assign(:current_organization, organization)
```
Comment (ruilopesm) [inline lib/atomic_web/live/activity_live/new.ex]: This assign is also not needed, by the reasons stated above.

File: lib/atomic_web/live/activity_live/form_component.html.heex [hunk 8]
```diff
@@ -1,47 +1,185 @@
...
+                          "hidden"
+                        end
+                      }>
+                        <svg class="h-5 w-5" viewBox="0 0 20 20" fill="currentColor" aria-hidden="true">
+                          <path fill-rule="evenodd" d="M16.704 4.153a.75.75 0 01.143 1.052l-8 10.5a.75.75 0 01-1.127.075l-4.5-4.5a.75.75 0 011.06-1.06l3.894 3.893 7.48-9.817a.75.75 0 011.05-.143z" clip-rule="evenodd" />
+                        </svg>
+                      </span>
+                    </li>
+                  <% end %>
+                </ul>
```
Comment (joaodiaslobo) [inline lib/atomic_web/live/activity_live/form_component.html.heex]: Can you make this a reusable component?

File: lib/atomic_web/live/activity_live/form_component.html.heex [hunk 9]
```diff
@@ -1,47 +1,185 @@
...
+                >
+                  <%= for speaker <- @speakers do %>
+                    <li class="flex items-center justify-between cursor-pointer select-none py-2 pr-3 pl-3 text-gray-900" role="option" phx-target={@myself} phx-click={JS.push("toggle_option", value: %{id: speaker.id})}>
+                      <span class="block truncate font-normal">
+                        <%= speaker.label %>
+                      </span>
+
+                      <span class={
+                        if speaker.selected do
+                          "inset-y-0 right-0 flex-row pr-4 text-indigo-600"
```
Comment (joaodiaslobo) [inline lib/atomic_web/live/activity_live/form_component.html.heex]:
```suggestion
                          "inset-y-0 right-0 flex-row pr-4 text-orange-500"
```

File: lib/atomic_web/components/forms.ex [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  def textarea(assigns) do
+    ~H"""
+    <div>
+      <span class="line-clamp-3 text-sm text-gray-700">
+        <span class="font-sm pl-1 text-base text-gray-800 sm:text-lg">
+          <%= @label %>
+        </span>
+        <%= textarea(@forms, @id, placeholder: @placeholder, rows: @rows, phx_debounce: "blur", class: "h-32 w-full rounded-md border-gray-300 font-medium shadow-sm focus:border-orange-500 focus:ring-orange-500 sm:text-sm") %>
```
Comment (joaodiaslobo) [inline lib/atomic_web/components/forms.ex]:
```suggestion
        <%= textarea(@forms, @id, placeholder: @placeholder, rows: @rows, phx_debounce: "blur", class: "h-32 w-full rounded-md border-gray-300 shadow-sm focus:border-orange-500 focus:ring-orange-500") %>
```

**Review by joaodiaslobo — CHANGES_REQUESTED** [review]

**Comment by diogogmatos** [comment]
[Add Activity Forms](https://trello.com/c/HcjeBXa6/2-add-activity-forms)

File: lib/atomic_web/live/activity_live/form_component.html.heex (hunk 9 above)
Comment (danielsp45) [reply lib/atomic_web/live/activity_live/form_component.html.heex]: This is a small icon. I think indigo looks better than orange

**Review by joaodiaslobo — CHANGES_REQUESTED** [review]
Can you make it so that, when the multiselect component focus is lost, it closes?

image

**Comment by joaodiaslobo** [comment]
Can you align the inputs in two columns with the same width? I don't really like them misaligned 😅

image

**Review by joaodiaslobo — APPROVED** [review]

**Pull request merged**